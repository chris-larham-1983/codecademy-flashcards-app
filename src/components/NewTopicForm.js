import React, { useState } from "react";
import { useDispatch } from 'react-redux';
import { useHistory } from "react-router-dom";
import { v4 as uuidv4 } from "uuid";
import ROUTES from "../app/routes";
import { ALL_ICONS } from "../data/icons";
import { addTopic } from '../features/topics/topicsSlice';

export default function NewTopicForm() {
  const [name, setName] = useState("");
  const [icon, setIcon] = useState("");
  const history = useHistory();
  const dispatch = useDispatch();

  const handleSubmit = (e) => {
    e.preventDefault();
    if (name.trim().length === 0) {
      document.getElementById('nameError').removeAttribute("hidden");
      return;
    }
    if(icon === "" || icon === "default") {
      document.getElementById('iconError').removeAttribute("hidden");
      return;
    }

    dispatch(addTopic({ id: uuidv4(), name: name, icon: icon }));
    history.push(ROUTES.topicsRoute());
  };

  return (
      <section>
        <form onSubmit={handleSubmit}>
          <h1 className="center">Create a new topic</h1>
          <div className="form-section">
            <input
                id="topic-name"
                type="text"
                value={name}
                onChange={(e) => {
                  setName(e.currentTarget.value);
                  if(e.currentTarget.value.trim() !== "" && !document.getElementById("nameError").getAttribute("hidden")) { //if the trimmed 'name' value does not equal an empty string AND the 'nameError' element does not have a 'hidden' attribute...
                    document.getElementById("nameError").setAttribute("hidden", "true"); //...set a 'hidden' attribute on the 'nameError' element
                  }
                }}
                placeholder="Topic Name"
            />
            <p id='nameError' hidden={true}>Please enter a topic name.</p>
            <select
                onChange={(e) => {
                  setIcon(e.currentTarget.value);
                  document.getElementById('iconError').setAttribute("hidden", "true"); //...set a 'hidden' attribute on the 'iconError' element
                }}
                required
                defaultValue="default"
                role="MENU"
            >
              <option value="default" disabled hidden>
                Choose an icon
              </option>
              {ALL_ICONS.map(({ name, url }) => (
                  <option key={url} value={url}>
                    {name}
                  </option>
              ))}
            </select>
            <p id='iconError' hidden={true}>Please select a topic icon.</p>
          </div>
          <button className="center">Add Topic</button>
        </form>
      </section>
  );
}
