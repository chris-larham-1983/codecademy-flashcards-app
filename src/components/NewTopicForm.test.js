import React from 'react';
import { Provider } from 'react-redux';
import TestRenderer from 'react-test-renderer';
import configureStore from 'redux-mock-store';
import NewTopicForm from './NewTopicForm';
import { render, fireEvent } from '@testing-library/react';
import { screen } from '@testing-library/dom';

const mockStore = configureStore([]);

describe("<NewTopicForm /> component", () => {
    let store;
    let component;

    beforeEach(() => {
        store = mockStore({
            topics: {
                '1': {
                    id: '1',
                    name: 'Topic 1',
                    icon: 'Icon URL 1',
                    quizIds: [],
                }
            }
        });

        component = TestRenderer.create(
            <Provider store={store}>
                <NewTopicForm />
            </Provider>
        );
    });

    describe("rendering", () => {
        it('should render the component without crashing', () => {
            expect(component.toJSON()).toBeTruthy();
        });

        it('should render with given state from the Redux store', () => {
            expect(component.toJSON()).toMatchSnapshot();
        });

        it('should render an h1 heading element', () => {
            expect(component.root.findByType('h1')).toBeTruthy();
            expect(component.root.findAllByProps({ className: "center" })[0].children).toStrictEqual(['Create a new topic']); //expect the first element whose className is 'center' (i.e. the h1 heading element) to display the text 'Create a new topic'
        });

        it('should render a form', () => {
            expect(component.root.findByType('form')).toBeTruthy();
        });

        it('should render an "Add Topic" button', () => {
            expect(component.root.findByType('button')).toBeTruthy();
            expect(component.root.findByType('button').children).toStrictEqual(['Add Topic']);
            expect(component.root.findAllByProps({ className: "center" }).length).toEqual(2); //expect there to be two elements (the h1 and the button) whose className is 'center'
        });

        it('should not initially display a paragraph that informs the User of the need to enter a topic name', () => {
            expect(component.root.findByProps({ id: "nameError" })).toBeTruthy();
            expect(component.root.findAllByProps({ hidden: true })[0]).toBeTruthy();
            expect(component.root.findAllByProps({ hidden: true })[0].children).toStrictEqual(['Please enter a topic name.']);
        });

        it('should not initially display a paragraph that informs the User of the need to select an icon', () => {
            expect(component.root.findByProps({ id: "iconError" })).toBeTruthy();
            expect(component.root.findAllByProps({ hidden: true })[2]).toBeTruthy();
            expect(component.root.findAllByProps({ hidden: true })[2].children).toStrictEqual(['Please select a topic icon.']);
        });

        describe('if a new topic form is submitted', () => {
            let store;
            let component;

            beforeEach(() => {
                store = mockStore({
                    topics: {
                        '1': {
                            id: '1',
                            name: 'Topic 1',
                            icon: 'Icon URL 1',
                            quizIds: [],
                        }
                    }
                });

                component = render(
                    <Provider store={store}>
                        <NewTopicForm />
                    </Provider>
                );
            });

            it('should display a descriptive paragraph to the User if no topic name is defined', () => {
                const addTopicButton = screen.getByText(/Add Topic/i);
                expect(addTopicButton).toBeInTheDocument(); //check that the button is in the document
                expect(screen.getByText(/Please enter a topic name./i)).not.toBeVisible(); //descriptive paragraph should initially be hidden
                fireEvent.click(addTopicButton); //click the button with no topic name entered
                expect(screen.getByText(/Please enter a topic name./i)).toBeVisible(); //expect the descriptive paragraph to be visible
            });

            it('should display a descriptive paragraph to the User if no topic icon is selected', () => {
                const addTopicButton = screen.getByText(/Add Topic/i);
                const topicNameInput = screen.getByPlaceholderText(/Topic Name/i);
                expect(topicNameInput.value).toBe(""); //expect the value of the topic name input element to be an empty string
                fireEvent.change(topicNameInput, { target: { value: 'New Topic Name' }}); //change the value of the topic name input element to 'New Topic Name'
                expect(topicNameInput.value).toBe('New Topic Name'); //expect the new value of the topic name input element to be 'New Topic Name'
                fireEvent.click(addTopicButton); //click the button with a topic name entered BUT no topic icon selected
                expect(screen.getByText(/Please enter a topic name./i)).not.toBeVisible(); //expect the descriptive paragraph related to the topic name not to be visible
                expect(screen.getByText(/Please select a topic icon./i)).toBeVisible(); //expect the descriptive paragraph related to the topic icon to be visible
            });
        });
        describe('when the User enters a topic name', () => {
            let store;
            let component;

            beforeEach(() => {
                store = mockStore({
                    topics: {
                        '1': {
                            id: '1',
                            name: 'Topic 1',
                            icon: 'Icon URL 1',
                            quizIds: [],
                        }
                    }
                });

                component = render(
                    <Provider store={store}>
                        <NewTopicForm />
                    </Provider>
                );
            });


            it('should hide the paragraph that instructs the User to enter a topic name', () => {
                const addTopicButton = screen.getByText(/Add Topic/i);
                const topicNameInput = screen.getByPlaceholderText(/Topic Name/i);
                expect(screen.getByText(/Please enter a topic name./i)).not.toBeVisible(); //descriptive paragraph should initially be hidden
                fireEvent.click(addTopicButton); //click the button with no topic name entered
                expect(screen.getByText(/Please enter a topic name./i)).toBeVisible(); //expect the descriptive paragraph to be visible
                fireEvent.change(topicNameInput, { target: { value: 'New Topic Name' }}); //change the value of the topic name input element to 'New Topic Name'
                expect(screen.getByText(/Please enter a topic name./i)).not.toBeVisible(); //expect the descriptive paragraph to no longer be visible
            });
        });

        describe('when the User selects a topic icon', () => {
            let store;
            let component;

            beforeEach(() => {
                store = mockStore({
                    topics: {
                        '1': {
                            id: '1',
                            name: 'Topic 1',
                            icon: 'Icon URL 1',
                            quizIds: [],
                        }
                    }
                });

                component = render(
                    <Provider store={store}>
                        <NewTopicForm />
                    </Provider>
                );
            });
            it('should hide the descriptive paragraph that instructs the User to select a topic icon', () => {
                const addTopicButton = screen.getByText(/Add Topic/i);
                const topicNameInput = screen.getByPlaceholderText(/Topic Name/i);
                const iconSelector = screen.getByRole('MENU');
                expect(iconSelector).toBeInTheDocument();
                fireEvent.change(topicNameInput, { target: { value: 'New Topic Name' }}); //change the value of the topic name input element to 'New Topic Name'
                fireEvent.click(addTopicButton); //click the button with a topic name entered BUT no topic icon selected
                expect(screen.getByText(/Please select a topic icon./i)).toBeVisible(); //expect the descriptive paragraph related to the topic icon to be visible
                iconSelector[2].selected = true; //set the iconSelector value to 'https://static-assets.codecademy.com/skillpaths/react-redux/redux-quiz-app/balloon.svg'
                fireEvent.change(iconSelector); //fire a change event on the iconSelector
                expect(screen.getByText(/Please select a topic icon./i)).not.toBeVisible(); //expect the descriptive paragraph related to the topic icon NOT to be visible
            });
        });
    });
});

