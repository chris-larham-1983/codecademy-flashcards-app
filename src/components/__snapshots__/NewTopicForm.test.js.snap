// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NewTopicForm /> component rendering should render with given state from the Redux store 1`] = `
<section>
  <form
    onSubmit={[Function]}
  >
    <h1
      className="center"
    >
      Create a new topic
    </h1>
    <div
      className="form-section"
    >
      <input
        id="topic-name"
        onChange={[Function]}
        placeholder="Topic Name"
        type="text"
        value=""
      />
      <p
        hidden={true}
        id="nameError"
      >
        Please enter a topic name.
      </p>
      <select
        defaultValue="default"
        onChange={[Function]}
        required={true}
        role="MENU"
      >
        <option
          disabled={true}
          hidden={true}
          value="default"
        >
          Choose an icon
        </option>
        <option
          value="https://static-assets.codecademy.com/skillpaths/react-redux/redux-quiz-app/book.svg"
        >
          Book
        </option>
        <option
          value="https://static-assets.codecademy.com/skillpaths/react-redux/redux-quiz-app/balloon.svg"
        >
          Balloon
        </option>
        <option
          value="https://static-assets.codecademy.com/skillpaths/react-redux/redux-quiz-app/bird.svg"
        >
          Bird
        </option>
        <option
          value="https://static-assets.codecademy.com/skillpaths/react-redux/redux-quiz-app/calendar.svg"
        >
          Calendar
        </option>
        <option
          value="https://static-assets.codecademy.com/skillpaths/react-redux/redux-quiz-app/clover.svg"
        >
          Clover
        </option>
        <option
          value="https://static-assets.codecademy.com/skillpaths/react-redux/redux-quiz-app/crayons.svg"
        >
          Crayons
        </option>
        <option
          value="https://static-assets.codecademy.com/skillpaths/react-redux/redux-quiz-app/data-flow.svg"
        >
          Data
        </option>
        <option
          value="https://static-assets.codecademy.com/skillpaths/react-redux/redux-quiz-app/fence.svg"
        >
          Fence
        </option>
        <option
          value="https://static-assets.codecademy.com/skillpaths/react-redux/redux-quiz-app/grill.svg"
        >
          Grill
        </option>
        <option
          value="https://static-assets.codecademy.com/skillpaths/react-redux/redux-quiz-app/hand-drill.svg"
        >
          Hand
        </option>
        <option
          value="https://static-assets.codecademy.com/skillpaths/react-redux/redux-quiz-app/hat.svg"
        >
          Hat
        </option>
        <option
          value="https://static-assets.codecademy.com/skillpaths/react-redux/redux-quiz-app/internet.svg"
        >
          Internet
        </option>
        <option
          value="https://static-assets.codecademy.com/skillpaths/react-redux/redux-quiz-app/ladybug.svg"
        >
          Ladybug
        </option>
        <option
          value="https://static-assets.codecademy.com/skillpaths/react-redux/redux-quiz-app/leaves.svg"
        >
          Leaves
        </option>
        <option
          value="https://static-assets.codecademy.com/skillpaths/react-redux/redux-quiz-app/medicine.svg"
        >
          Medicine
        </option>
        <option
          value="https://static-assets.codecademy.com/skillpaths/react-redux/redux-quiz-app/nest.svg"
        >
          Nest
        </option>
        <option
          value="https://static-assets.codecademy.com/skillpaths/react-redux/redux-quiz-app/shuttlecock.svg"
        >
          Shuttlecock
        </option>
        <option
          value="https://static-assets.codecademy.com/skillpaths/react-redux/redux-quiz-app/spade.svg"
        >
          Spade
        </option>
        <option
          value="https://static-assets.codecademy.com/skillpaths/react-redux/redux-quiz-app/statistics.svg"
        >
          Statistics
        </option>
        <option
          value="https://static-assets.codecademy.com/skillpaths/react-redux/redux-quiz-app/sun.svg"
        >
          Sun
        </option>
        <option
          value="https://static-assets.codecademy.com/skillpaths/react-redux/redux-quiz-app/tree.svg"
        >
          Tree
        </option>
      </select>
      <p
        hidden={true}
        id="iconError"
      >
        Please select a topic icon.
      </p>
    </div>
    <button
      className="center"
    >
      Add Topic
    </button>
  </form>
</section>
`;
